cmd/util/version.go

```diff
@@ -65,23 +69,82 @@ func GetAllVersions(ctx context.Context) (Versions, error) {
 
 var printMessage *string = nil
 
+const updateStatePath string = "update.json"
+const updateFrequency time.Duration = 24 * time.Hour
+
 // StartUpdateCheck is a Cobra pre run hook to run an update check in the
 // background. There should be no output if the check fails or the context is
 // cancelled before the check can complete.
 func StartUpdateCheck(cmd *cobra.Command, args []string) {
-	go func(ctx context.Context) {
-		if skip, err := config.Get[bool](types.SkipUpdateCheck); skip || err != nil {
-			log.Ctx(ctx).Debug().Err(err).Bool(types.SkipUpdateCheck, skip).Msg("Skipping update check")
-			return
-		}
+	if skip, err := config.Get[bool](types.SkipUpdateCheck); skip || err != nil {
+		log.Ctx(cmd.Context()).Debug().Err(err).Bool(types.SkipUpdateCheck, skip).Msg("Skipping update check due to config")
+		return
+	}
 
+	lastCheck, err := readLastCheckTime(cmd.Context())
+	if err == nil && time.Since(lastCheck) < updateFrequency {
+		log.Ctx(cmd.Context()).Debug().Time("LastCheck", lastCheck).Msg("Skipping update check as there was a recent enough check")
+		return
+	} else if err != nil {
+		log.Ctx(cmd.Context()).Warn().Err(err).Msg("Error reading update check state – will perform check anyway")
+	}
+
+	go func(ctx context.Context) {
 		versions, err := GetAllVersions(ctx)
 		if err == nil {
 			printMessage = &versions.UpdateMessage
+			err = writeNewLastCheckTime(ctx)
+			log.Ctx(ctx).Debug().Err(err).Str("LatestVersion", versions.LatestVersion.GitVersion).Msg("Performed update check")
 		}
 	}(cmd.Context())
```

Comment:
Is this (cmd.Context()) the ctx with the value attached in root.go?

> ctx = cmd.Context()
> ... 
> ctx = context.WithValue(ctx, util.FSRepoKey, fsRepo)

It's kinda confusing that this works calling .Context()

Reply:
Yes, there is a `cmd.SetContext()` call that overlays the new context into the command before the end of the pre-run hook.